src/visualisation/panel.cc

```diff
@@ -230,97 +241,127 @@ void VisualisationPanel::refreshVisualisation() {
   initOptionsPanel();
 
-  QWidget* root_widget = new QWidget(this);
-  this->setCentralWidget(root_widget);
-
-  splitter_ = new QSplitter(Qt::Horizontal);
-  splitter_->addWidget(minimap_);
-  splitter_->addWidget(visualisation_);
-  splitter_->setStretchFactor(1, 5);
-
-  layout_ = new QHBoxLayout;
-  layout_->setContentsMargins(0, 0, 0, 0);
-  layout_->addWidget(splitter_, 8);
-  layout_->addLayout(options_layout_, 0);
-  root_widget->setLayout(layout_);
+  setCentralWidget(visualisation_root_);
+  setDockNestingEnabled(true);
+
+  node_tree_dock_ = new QDockWidget;
+  new ui::DockWidgetVisibilityGuard(node_tree_dock_);
+  node_tree_dock_->setWindowTitle("Node tree");
+  QSharedPointer<QItemSelectionModel> new_selection_model(
+            new QItemSelectionModel(data_model_.data()));
+  node_tree_widget_ = new ui::NodeTreeWidget(main_window_, data_model_,
+      new_selection_model);
+  node_tree_dock_->setWidget(node_tree_widget_);
+  node_tree_dock_->setContextMenuPolicy(Qt::PreventContextMenu);
+  node_tree_dock_->setAllowedAreas(
+      Qt::LeftDockWidgetArea | Qt::RightDockWidgetArea);
+  node_tree_dock_->hide();
+  addDockWidget(Qt::LeftDockWidgetArea, node_tree_dock_);
+
+  minimap_dock_ = new QDockWidget;
+  new ui::DockWidgetVisibilityGuard(minimap_dock_);
+  minimap_dock_->setWindowTitle("Minimap");
+  minimap_dock_->setWidget(minimap_);
+  minimap_dock_->setContextMenuPolicy(Qt::PreventContextMenu);
+  minimap_dock_->setAllowedAreas(
+      Qt::LeftDockWidgetArea | Qt::RightDockWidgetArea);
+  ui::MainWindowWithDetachableDockWidgets::splitDockWidget2(this,
+      node_tree_dock_, minimap_dock_, Qt::Horizontal);
+
+  connect(show_node_tree_act_, &QAction::toggled,
+      node_tree_dock_, &QDockWidget::setVisible);
+  connect(show_minimap_act_, &QAction::toggled,
+      minimap_dock_, &QDockWidget::setVisible);
 }
 
-QBoxLayout* VisualisationPanel::prepareVisualisationOptions() {
-  QBoxLayout *child_layout = new QVBoxLayout;
-  child_layout->setContentsMargins(0, 0, 0, 0);
-  visualisation_->prepareOptionsPanel(child_layout);
-  return child_layout;
+void VisualisationPanel::prepareVisualisationOptions() {
+  visualisation_->prepareOptions(visualisation_root_);
 }
 
 void VisualisationPanel::initOptionsPanel() {
-  options_layout_ = new QVBoxLayout;
-
+  /////////////////////////////////////
+  // Node tree / minimap
+  show_node_tree_act_ = new QAction(QIcon(":/images/show_node_tree.png"),
+      tr("&Node tree"), this);
+  show_node_tree_act_->setToolTip(tr("Node tree"));
+  show_node_tree_act_->setEnabled(true);
+  show_node_tree_act_->setCheckable(true);
+  show_node_tree_act_->setChecked(false);
+  show_minimap_act_ = new QAction(QIcon(":/images/show_minimap.png"),
+      tr("&Minimap"), this);
+  show_minimap_act_->setToolTip(tr("Minimap"));
+  show_minimap_act_->setEnabled(true);
+  show_minimap_act_->setCheckable(true);
+  show_minimap_act_->setChecked(true);
+
+  tools_tool_bar_ = new QToolBar(tr("Tools"));
```

Comment:
Similar thing here - `addToolBar` doesn't mention that it takes ownership of widget.